packages/@react-spectrum/tag/src/TagGroup.tsx

```diff
@@ -10,52 +10,154 @@
+        let end = direction === 'ltr' ? 'right' : 'left';
+        let containerEnd = currDomRef.getBoundingClientRect()[end];
+        let lastTagEnd = tags[index - 1]?.getBoundingClientRect()[end];
+        let availableWidth = containerEnd - lastTagEnd;
+        for (let tagWidth of tagWidths.reverse()) {
+          if (availableWidth >= buttonWidth || index <= 1 || index >= state.collection.size) {
+            break;
+          }
+          availableWidth += tagWidth;
+          index--;
+        }
+        return {visibleTagCount: index, showCollapseButton: index < state.collection.size};
+      };
+    
+      setTagState(function *() {
+        // Update to show all items.
+        yield {visibleTagCount: state.collection.size, showCollapseButton: true};
+
+        // Measure, and update to show the items until maxRows is reached.
+        yield computeVisibleTagCount();
+      });
+    }
+  }, [direction, domRef, state.collection.size, maxRows, setTagState]);
+
+  useResizeObserver({ref: domRef, onResize: updateVisibleTagCount});
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  useLayoutEffect(updateVisibleTagCount, [children]);
+
+  let visibleTags = [...state.collection];
+  if (maxRows != null && isCollapsed) {
+    visibleTags = visibleTags.slice(0, tagState.visibleTagCount);
+  }
+
+  let handlePressCollapse = () => {
+    // Prevents button from losing focus if focusedKey got collapsed.
+    state.selectionManager.setFocusedKey(null);
+    setIsCollapsed(prevCollapsed => !prevCollapsed);
+  };
+
   return (
-    <div
-      {...mergeProps(styleProps, tagGroupProps)}
-      className={
-        classNames(
-          styles,
-          'spectrum-Tags',
-          styleProps.className
-        )
-      }
-      role={state.collection.size ? 'grid' : null}
-      ref={domRef}>
-      {[...state.collection].map(item => (
-        <Tag
-          {...item.props}
-          key={item.key}
-          item={item}
-          state={state}
-          allowsRemoving={allowsRemoving}
-          onRemove={onRemove}>
-          {item.rendered}
-        </Tag>
-      ))}
-    </div>
+    <FocusScope>
+      <div
+        {...mergeProps(styleProps, tagGroupProps)}
+        className={
+          classNames(
+            styles,
+            'spectrum-Tags',
+            styleProps.className
+          )
+        }
+        role={state.collection.size ? 'grid' : null}
+        ref={domRef}>
+        {visibleTags.map(item => (
+          <Tag
+            {...item.props}
+            key={item.key}
+            item={item}
+            state={state}
+            allowsRemoving={allowsRemoving}
+            onRemove={onRemove}>
+            {item.rendered}
+          </Tag>
+        ))}
+        {tagState.showCollapseButton &&
+          <ActionButton isQuiet onPress={handlePressCollapse}>
+            {isCollapsed ? `Show all (${state.collection.size})` : 'Show less '}
+          </ActionButton>
+        }
```

Comment:
I think this will be invalid aria because it's a button directly inside a `role="grid"` without any `row` or `gridcell` wrappers. Also, since you tab to these rather than arrow key navigating, maybe they should actually be outside the grid? Would that break the layout?

Reply:
Yeah, I wanted to include it outside grid, but couldn't figure out how to get it on the last row of tags. I'll keep looking.

Reply:
Managed to get it outside of the grid and keep the layout correct.